bliss/models/fnp.py

```diff
@@ -0,0 +1,464 @@
+        input_list = [yR_encoded]
+        if self.use_x:
+            input_list.append(XR.unsqueeze(0))
+
+        ## If we look at differences in U values, we need to increase the dimension
+        ## (each representative member looks different to each dependent member)
+        if self.use_u_diff:
+            u_diff = u.unsqueeze(1) - uR.unsqueeze(0)
+            for i, x in enumerate(input_list):
+                input_list[i] = x.unsqueeze(1)
+            input_list.append(u_diff.unsqueeze(0))
+
+        rep_R = self.f(self.cat(*input_list))
+        return rep_R
+
+
+# **************************
+# Representation Poolers
+# **************************
+class AveragePooler(nn.Module):
+    """
+    Pools together representations by taking a sum.
+    """
+
+    def __init__(
+        self,
+        dim_z,
+    ):
+        super().__init__()
+        self.dim_z = dim_z
+
+        # normalizes the graph such that inner products correspond to averages of the parents
+        self.norm_graph = lambda x: x / (torch.sum(x, 1, keepdim=True) + 1e-8)
+
+    def forward(self, rep_R, GA):
+        W = self.norm_graph(GA)
+        pz_all = torch.matmul(W, rep_R)
+        return pz_all
+
+
+class SetPooler(nn.Module):
+    """
+    Pools together representations using a set transformer architecture.
+    See https://arxiv.org/abs/1810.00825.
+    """
+
+    def __init__(
+        self,
+        dim_rep_R,
+        dim_z,
+        pooling_layers,
+        set_transformer,
+        st_numheads,
+    ):
+        super().__init__()
+        self.dim_rep_R = dim_rep_R
+        self.dim_z = dim_z
+        self.pooling_layers = pooling_layers
+        self.set_transformer = set_transformer
+        self.st_numheads = st_numheads
+
+        # normalizes the graph such that inner products correspond to averages of the parents
+        self.norm_graph = lambda x: x / (torch.sum(x, 1, keepdim=True) + 1e-8)
+
+        if not self.set_transformer:
+            self.pool_net = MLP(self.dim_rep_R, self.pooling_layers, 2 * self.dim_z)
+            self.settrans = None
+        else:
+            self.pool_net = None
+            dim_in = self.dim_rep_R
+            sabs = [SAB(dim_in, dim_in, nh) for nh in self.st_numheads]
+            self.settrans = nn.Sequential(
+                *sabs,
+                PMA(dim_in, 2, 1, squeeze_out=True),
+                MLP(dim_in, self.pooling_layers, 2 * self.dim_z),
+            )
+
+    def forward(self, rep_R, GA):
+        if not self.set_transformer:
+            rep_pooled = GA.unsqueeze(0).unsqueeze(-1).mul(rep_R).sum(2)
+            pz_all = self.pool_net(rep_pooled)
+        else:
+            pz_all = self.settrans(GA.unsqueeze(0).unsqueeze(-1).mul(rep_R))
+        return pz_all
+
+
+# *********************
+# Set Transformer
+# *********************
+# This implementation is taken from https://github.com/juho-lee/set_transformer/,
+# which comes with the MIT license.
+
+
+class MAB(nn.Module):
```

Comment:
Hi Derek, Would you add some brief docstrings to these classes, and re-run the makefile in the docs directory?

Reply:
I added the docstrings, but I don't know what I am supposed to do to regenerate the documentation

Reply:
I think it's @zzhaozhe-profolio said to type `make github` from within the `docs/docstr` directory. It's also fine to leave it for another time

Reply:
I see, I figured it out after poking around a bit. I also had to update one of the config files to have it scan the fnp.py file. 

I made the changes in the last commit on this branch, and added a small README.md file to the docs to show what to do in the future.